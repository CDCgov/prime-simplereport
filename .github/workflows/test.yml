name: Test

# Run this on pushes to any branch that change a backend file or the workflow definition
on:
  workflow_dispatch: # because sometimes you just want to force a branch to have tests run
  pull_request:
    branches:
      - "**"
  push:
    branches:
      - main

env:
  NODE_VERSION: 14
  JAVA_VERSION: 11
  PROJECT_ROOT: /home/runner/work/prime-simplereport/prime-simplereport # Hardcoding this here because env.WORKSPACE_ROOT gets overridden by one of the steps downstream. We only need this for Sonar.

defaults:
  run:
    working-directory: backend

jobs:
  test:
    runs-on: ubuntu-latest
    services:
      test-db:
        image: postgres:11-alpine
        env:
          POSTGRES_PASSWORD: this_is_a_super_secure_admin_password
        # Set health checks to wait until postgres has started
        options: >-
          --health-cmd pg_isready
          --health-interval 10s
          --health-timeout 5s
          --health-retries 5
        ports:
          - 5432:5432
    steps:
      - uses: actions/checkout@v2
        with:
          fetch-depth: 0 # Disable shallow clones so Sonar can have all the data
      - name: Set up JDK ${{env.JAVA_VERSION}}
        uses: actions/setup-java@v1
        with:
          java-version: ${{env.JAVA_VERSION}}
      - name: Cache Java Dependencies
        uses: actions/cache@v2
        with:
          path: |
            ~/.gradle/caches
            ~/.gradle/wrapper
          key: gradle-${{ hashFiles('*.gradle', 'gradle/dependency-locks/*') }}
      - name: Set up postgresql sidecar
        env:
          PGHOST: localhost
          PGPASSFILE: db-setup/.pgpass_ci
          POSTGRES_USER: postgres
          POSTGRES_DB: postgres
          LIB_DIR: db-setup
        run: |
          chmod 0600 $PGPASSFILE
          db-setup/create-db.sh
      - name: Run Java Tests
        env:
          OKTA_TESTING_DISABLEHTTPS: true
          TWILIO_ACCOUNT_SID: ${{secrets.TWILIO_TEST_ACCOUNT_SID }}
          TWILIO_AUTH_TOKEN: ${{secrets.TWILIO_TEST_AUTH_TOKEN }}
        run: ./gradlew jacocoTestReport -PtestDbPort=5432
      - name: Archive Test Results
        uses: actions/upload-artifact@v2
        if: failure()
        with:
          name: test-report
          path: backend/build/reports/tests/test
          retention-days: 7
      - name: Use Node.js ${{env.NODE_VERSION}}
        uses: actions/setup-node@v2.1.5
        with:
          node-version: ${{env.NODE_VERSION}}
      - name: Cache yarn
        uses: actions/cache@v2
        with:
          path: ~/.cache/yarn
          key: ${{ runner.os }}-yarn-${{ hashFiles('**/yarn.lock') }}
          restore-keys: |
            ${{ runner.os }}-yarn-
      - name: Node setup
        working-directory: ./frontend
        run: yarn install --prefer-offline
      - name: Test
        working-directory: ./frontend
        run: yarn test:ci
      - name: "Function Apps: npm install"
        working-directory: ./ops/services/app_functions/report_stream_batched_publisher/functions
        run: npm ci
      - name: "Function Apps: Test"
        working-directory: ./ops/services/app_functions/report_stream_batched_publisher/functions
        run: npm run coverage
      - name: Sonar analysis
        if: ${{ github.actor != 'dependabot[bot]' }}
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }} # Needed to get PR information, if any
          SONAR_TOKEN: ${{ secrets.SONAR_TOKEN }}
        run: ./gradlew sonarqube -Dsonar.projectBaseDir=${{ env.PROJECT_ROOT }} --info
  build-jar:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - name: Set up JDK ${{env.JAVA_VERSION}}
        uses: actions/setup-java@v1
        with:
          java-version: ${{env.JAVA_VERSION}}
      - name: Cache Java Dependencies
        uses: actions/cache@v2
        with:
          path: |
            ~/.gradle/caches
            ~/.gradle/wrapper
          key: gradle-${{ hashFiles('*.gradle', 'gradle/dependency-locks/*') }}
      - name: Run Java Build
        run: ./gradlew assemble
  backend-lint:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - name: Set up JDK ${{env.JAVA_VERSION}}
        uses: actions/setup-java@v1
        with:
          java-version: ${{env.JAVA_VERSION}}
      - name: Cache Java Dependencies
        uses: actions/cache@v2
        with:
          path: |
            ~/.gradle/caches
            ~/.gradle/wrapper
          key: gradle-${{ hashFiles('*.gradle', 'gradle/dependency-locks/*') }}
      - name: Run Style checks
        run: ./gradlew checkstyleMain checkstyleTest spotlessCheck --continue
  frontend-lint-and-build:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - name: Use Node.js ${{env.NODE_VERSION}}
        uses: actions/setup-node@v2.1.5
        with:
          node-version: ${{env.NODE_VERSION}}
      - name: Cache yarn
        uses: actions/cache@v2
        with:
          path: ~/.cache/yarn
          key: ${{ runner.os }}-yarn-${{ hashFiles('**/yarn.lock') }}
          restore-keys: |
            ${{ runner.os }}-yarn-
      - name: Node setup
        working-directory: ./frontend
        run: yarn
      - name: Lint front-end
        working-directory: ./frontend
        run: yarn lint
      - name: Build release
        working-directory: ./frontend
        run: yarn build
