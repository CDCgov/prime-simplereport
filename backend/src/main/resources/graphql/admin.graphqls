# Queries and mutations that are used only by site admins
# All queries and mutations in this file implicitly require site admin access,
# which is enforced in the API not in the schema validator.
extend type Query {
  organizations(identityVerified: Boolean): [Organization!]!
  pendingOrganizations: [PendingOrganization!]!
}
extend type Mutation {
  uploadPatients(patientList: Upload!): String
  resendToReportStream(testEventIds: [ID!]!): Boolean
  createDeviceType(input: CreateDeviceType!): DeviceType
  updateDeviceType(input: UpdateDeviceType!): DeviceType
  addUser( # not actually used in the admin console (currently)
    name: NameInput
    firstName: String
    middleName: String
    lastName: String
    suffix: String
    email: String!
    organizationExternalId: String!
    role: Role!
  ): User
  createOrganization(
    name: String!
    type: String!
    externalId: String!
    testingFacilityName: String!
    cliaNumber: String
    street: String!
    streetTwo: String
    city: String
    county: String
      @deprecated(reason: "county is derived from smarty streets lookup")
    state: String!
    zipCode: String!
    phone: String
    email: String
    orderingProviderName: NameInput
    orderingProviderFirstName: String
    orderingProviderMiddleName: String
    orderingProviderLastName: String
    orderingProviderSuffix: String
    orderingProviderNPI: String
    orderingProviderStreet: String
    orderingProviderStreetTwo: String
    orderingProviderCity: String
    orderingProviderCounty: String
      @deprecated(reason: "county is derived from smarty streets lookup")
    orderingProviderState: String
    orderingProviderZipCode: String
    orderingProviderPhone: String
    deviceTypes: [String]!
    defaultDevice: String!
    adminName: NameInput
    adminFirstName: String
    adminMiddleName: String
    adminLastName: String
    adminSuffix: String
    adminEmail: String!
  ): Organization
  setOrganizationIdentityVerified(
    externalId: String!
    verified: Boolean!
  ): Boolean
  editPendingOrganization(
    orgExternalId: String!
    name: String
    adminFirstName: String
    adminLastName: String
    adminEmail: String
    adminPhone: String
  ): String
  markPendingOrganizationAsDeleted(
    orgExternalId: String!
    deleted: Boolean!
  ): String
  createOrganizationRegistrationLink(
    organizationExternalId: String!
    link: String!
  ): String
  createFacilityRegistrationLink(
    organizationExternalId: String!
    facilityId: ID!
    link: String!
  ): String
  updateRegistrationLink(link: String!, newLink: String!): String
  setRegistrationLinkIsDeleted(link: String, deleted: Boolean!): String
  adminUpdateOrganization(name: String!, type: String!): String
  markFacilityAsDeleted(facilityId: ID!, deleted: Boolean!): String
}
