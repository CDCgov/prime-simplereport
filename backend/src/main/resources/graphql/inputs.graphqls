input ProviderInput {
    fullName:NameInput
    firstName: String
    middleName: String
    lastName: String
    suffix: String
    npi: String
    street:String
    streetTwo:String
    city: String
    state: String
    zipCode: String
    phone: String
}