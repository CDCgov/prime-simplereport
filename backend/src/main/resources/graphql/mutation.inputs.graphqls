# Contains input definitions for mutations
input ProviderInput {
    firstName: String
    middleName: String
    lastName: String
    suffix: String
    npi: String
    street:String
    streetTwo:String
    city: String
    county: String
    state: String
    zipCode: String
    phone: String
}

input FacilityAddressInput{
    street: String!
    streetTwo: String
    city: String
    state: String!
    zipCode: String!
}

input UpdateFacilityInput {
    facilityId: ID!
    facilityName: String!
    cliaNumber: String
    address: FacilityAddressInput!
    phone: String
    email: String
    orderingProvider: ProviderInput
    deviceIds: [ID]!
}

input AddFacilityInput {
    facilityName: String!
    cliaNumber: String
    address: FacilityAddressInput!
    phone: String
    email: String
    orderingProvider: ProviderInput
    deviceIds: [ID]!
}