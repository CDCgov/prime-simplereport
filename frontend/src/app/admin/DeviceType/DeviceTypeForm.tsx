import React, { useState } from "react";

import Button from "../../commonComponents/Button/Button";
import TextInput from "../../commonComponents/TextInput";

import { Device } from "./DeviceTypeFormContainer";

interface Props {
  saveDeviceType: (device: Device) => void;
}

const DeviceTypeForm: React.FC<Props> = ({ saveDeviceType }) => {
  const [device, updateDevice] = useState<Device>({
    name: "",
    manufacturer: "",
    model: "",
    loincCode: "",
    swabType: "",
  });
  const [formChanged, updateFormChanged] = useState<boolean>(false);
  const onChange = (
    e: React.ChangeEvent<HTMLInputElement | HTMLSelectElement>
  ) => {
    updateDevice({ ...device, [e.target.name]: e.target.value });
    updateFormChanged(true);
  };
  return (
    <main className="prime-home">
      <div className="grid-container">
        <div className="grid-row">
          <div className="prime-container card-container">
            <div className="usa-card__header">
              <div>
                <h2>Device Type</h2>
              </div>
              <div
                style={{
                  display: "flex",
                  justifyContent: "center",
                  alignItems: "center",
                }}
              >
                <Button
                  type="button"
                  onClick={() => saveDeviceType(device)}
                  label="Save Changes"
                  disabled={!formChanged}
                />
              </div>
            </div>
            <div className="usa-card__body margin-top-1">
              <div>
                <div className="usa-alert usa-alert--warning">
                  <div className="usa-alert__body">
                    <h4 className="usa-alert__heading">Reminder</h4>
                    <p className="usa-alert__text">
                      Notify ReportStream of devices added in{" "}
                      <a
                        href="https://usds.slack.com/archives/C024MGSJZ38"
                        target="_blank"
                      >
                        #prime-reportstream
                      </a>
                    </p>
                    <p>
                      Device details can be found by downloading the excel file
                      published at{" "}
                      <a
                        href="https://www.cdc.gov/csels/dls/sars-cov-2-livd-codes.html"
                        target="_blank"
                      >
                        https://www.cdc.gov/csels/dls/sars-cov-2-livd-codes.html
                      </a>
                    </p>
                  </div>
                </div>
              </div>
              <div className="grid-row grid-gap">
                <div className="tablet:grid-col">
                  <TextInput
                    label="Name"
                    name="name"
                    value={device.name}
                    onChange={onChange}
                    required
                  />
                </div>
                <div className="tablet:grid-col">
                  <TextInput
                    label="Manufacturer"
                    name="manufacturer"
                    value={device.manufacturer}
                    onChange={onChange}
                    required
                  />
                </div>
                <div className="tablet:grid-col">
                  <TextInput
                    label="Model"
                    name="model"
                    value={device.model}
                    onChange={onChange}
                    required
                  />
                </div>
                <div className="tablet:grid-col">
                  <TextInput
                    label="Loinc Code"
                    name="loincCode"
                    value={device.loincCode}
                    onChange={onChange}
                    required
                  />
                </div>
              </div>
              <div className="grid-row grid-gap">
                <div className="tablet:grid-col">
                  <TextInput
                    label="SNOMED code of Swab Type"
                    name="swabType"
                    value={device.swabType}
                    onChange={onChange}
                    required
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </main>
  );
};

export default DeviceTypeForm;
