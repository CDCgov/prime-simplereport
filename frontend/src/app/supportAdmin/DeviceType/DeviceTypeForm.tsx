import React, { useState } from "react";

import Button from "../../commonComponents/Button/Button";
import TextInput from "../../commonComponents/TextInput";
import MultiSelect from "../../commonComponents/MultiSelect/MultiSelect";
import { MultiSelectDropdownOption } from "../../commonComponents/MultiSelect/MultiSelectDropdown/MultiSelectDropdown";

import { Device } from "./DeviceTypeFormContainer";
import DeviceTypeReminderMessage from "./DeviceTypeReminderMessage";

interface Props {
  saveDeviceType: (device: Device) => void;
  swabOptions: Array<MultiSelectDropdownOption>;
  supportedDiseaseOptions: Array<MultiSelectDropdownOption>;
}

const DeviceTypeForm: React.FC<Props> = ({
  saveDeviceType,
  swabOptions,
  supportedDiseaseOptions,
}) => {
  const [device, updateDevice] = useState<Device>({
    name: "",
    manufacturer: "",
    model: "",
    loincCode: "",
    swabTypes: [],
    supportedDiseases: [],
  });
  const [formChanged, updateFormChanged] = useState<boolean>(false);

  function updateDeviceAttribute(name: string, value: any) {
    updateDevice({ ...device, [name]: value });
    updateFormChanged(true);
  }

  const onChange = (
    e: React.ChangeEvent<HTMLInputElement | HTMLSelectElement>
  ) => {
    updateDeviceAttribute(e.target.name, e.target.value);
  };
  return (
    <main className="prime-home">
      <div className="grid-container">
        <div className="grid-row">
          <div className="prime-container card-container">
            <div className="usa-card__header">
              <div>
                <h2>Device type</h2>
              </div>
              <div
                style={{
                  display: "flex",
                  justifyContent: "center",
                  alignItems: "center",
                }}
              >
                <Button
                  type="button"
                  onClick={() => saveDeviceType(device)}
                  label="Save changes"
                  disabled={!formChanged}
                />
              </div>
            </div>
            <div className="usa-card__body margin-top-1">
              <DeviceTypeReminderMessage />
              <div className="grid-row grid-gap">
                <div className="tablet:grid-col">
                  <TextInput
                    label="Device name"
                    name="name"
                    value={device.name}
                    onChange={onChange}
                    required
                  />
                </div>
                <div className="tablet:grid-col">
                  <TextInput
                    label="Manufacturer"
                    name="manufacturer"
                    value={device.manufacturer}
                    onChange={onChange}
                    required
                  />
                </div>
                <div className="tablet:grid-col">
                  <TextInput
                    label="Model"
                    name="model"
                    value={device.model}
                    onChange={onChange}
                    required
                  />
                </div>
                <div className="tablet:grid-col">
                  <TextInput
                    label="LOINC code"
                    name="loincCode"
                    value={device.loincCode}
                    onChange={onChange}
                    required
                  />
                </div>
              </div>
              <div className="grid-row grid-gap">
                <div className="tablet:grid-col">
                  <MultiSelect
                    label="SNOMED code for swab type(s)"
                    name="swabTypes"
                    onChange={(swabTypes) => {
                      updateDeviceAttribute("swabTypes", swabTypes);
                    }}
                    options={swabOptions}
                    required
                  />
                </div>
              </div>
              <div className="grid-row grid-gap">
                <div
                  className="tablet:grid-col"
                  style={{ marginBottom: "56px" }}
                >
                  <MultiSelect
                    label="Supported diseases"
                    name="supportedDiseases"
                    onChange={(supportedDiseases) => {
                      updateDeviceAttribute(
                        "supportedDiseases",
                        supportedDiseases
                      );
                    }}
                    options={supportedDiseaseOptions}
                    required
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </main>
  );
};

export default DeviceTypeForm;
