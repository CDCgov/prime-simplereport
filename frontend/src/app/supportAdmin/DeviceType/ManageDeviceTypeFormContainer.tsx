import { useState } from "react";
import { Navigate } from "react-router-dom";

import {
  UpdateDeviceType,
  useGetDeviceTypeListQuery,
  useGetSpecimenTypesQuery,
  useUpdateDeviceTypeMutation,
  DeviceType,
} from "../../../generated/graphql";
import { LoadingCard } from "../../commonComponents/LoadingCard/LoadingCard";
import { showNotification } from "../../utils";
import Alert from "../../commonComponents/Alert";

import ManageDevicesForm from "./ManageDevicesForm";

const ManageDeviceTypeFormContainer = () => {
  const [submitted, setSubmitted] = useState(false);
  const [updateDeviceType] = useUpdateDeviceTypeMutation();
  const { data: specimenTypesResults } = useGetSpecimenTypesQuery({
    fetchPolicy: "no-cache",
  });
  const { data: deviceTypeResults } = useGetDeviceTypeListQuery({
    fetchPolicy: "no-cache",
  });

  const saveDeviceType = (device: UpdateDeviceType) => {
    updateDeviceType({
      variables: device,
      fetchPolicy: "no-cache",
    }).then(() => {
      const alert = (
        <Alert
          type="success"
          title="Created Device"
          body="The device has been created"
        />
      );
      showNotification(alert);
      setSubmitted(true);
    });
  };

  if (submitted) {
    return <Navigate to="/admin" />;
  }

  if (deviceTypeResults && specimenTypesResults) {
    const swabOptions = Array.from(
      specimenTypesResults.specimenTypes.map((type) => ({
        label: `${type?.name} (${type?.typeCode})`,
        value: type?.internalId,
      }))
    );

    const devices = Array.from(
      deviceTypeResults.deviceTypes.map(
        (devicesTypes) => devicesTypes as DeviceType
      )
    );

    return (
      <ManageDevicesForm
        updateDeviceType={saveDeviceType}
        swabOptions={swabOptions}
        devices={devices}
      />
    );
  } else {
    return <LoadingCard message="Loading" />;
  }
};

export default ManageDeviceTypeFormContainer;
