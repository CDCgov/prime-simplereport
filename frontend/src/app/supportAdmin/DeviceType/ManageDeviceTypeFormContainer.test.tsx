import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { SpecimenType, DeviceType } from "../../../generated/graphql";

import ManageDeviceTypeFormContainer from "./ManageDeviceTypeFormContainer";
import { addValue } from "./DeviceTypeForm.test";

const mockUpdateDeviceType = jest.fn();

jest.mock("../../../generated/graphql", () => {
  return {
    useUpdateDeviceTypeMutation: () => [
      (options: any) => {
        mockUpdateDeviceType(options);
        return Promise.resolve();
      },
    ],
    useGetSpecimenTypesQuery: () => {
      return {
        data: {
          specimenTypes: [
            {
              internalId: "123",
              name: "nose",
              typeCode: "123",
            },
            {
              internalId: "456",
              name: "eyes",
              typeCode: "456",
            },
            {
              internalId: "789",
              name: "mouth",
              typeCode: "789",
            },
          ] as SpecimenType[],
        },
      };
    },
    useGetDeviceTypeListQuery: () => {
      return {
        data: {
          deviceTypes: [
            {
              internalId: "abc1",
              name: "Tesla Emitter",
              model: "Model A",
              manufacturer: "Celoxitin",
              loincCode: "1234-1",
              swabTypes: [
                { internalId: "123", name: "nose", typeCode: "n123" },
              ],
            },
            {
              internalId: "abc2",
              name: "Fission Energizer",
              model: "Model B",
              manufacturer: "Curentz",
              loincCode: "1234-2",
              swabTypes: [{ internalId: "456", name: "eye", typeCode: "e456" }],
            },
            {
              internalId: "abc3",
              name: "Covalent Observer",
              model: "Model C",
              manufacturer: "Vitamin Tox",
              loincCode: "1234-3",
              swabTypes: [
                { internalId: "789", name: "mouth", typeCode: "m789" },
              ],
            },
          ] as DeviceType[],
        },
      };
    },
  };
});

jest.mock("react-router-dom", () => {
  const original = jest.requireActual("react-router-dom");
  return {
    ...original,
    Navigate: (props: any) => `Redirected to ${props.to}`,
  };
});

describe("ManageDeviceTypeFormContainer", () => {
  beforeEach(() => {
    render(<ManageDeviceTypeFormContainer />);
  });

  it("should show the device type form", async () => {
    expect(await screen.findByText("Manage devices")).toBeInTheDocument();
  });

  it("should update the selected device", async () => {
    await new Promise((resolve) => setTimeout(resolve, 0));

    userEvent.selectOptions(
      screen.getByLabelText("Device name", { exact: false }),
      "Covalent Observer"
    );

    await new Promise((resolve) => setTimeout(resolve, 0));

    addValue("Manufacturer", " LLC");

    addValue("Model", "D");

    userEvent.click(screen.getByText("Save changes"));

    expect(mockUpdateDeviceType).toBeCalledTimes(1);
    expect(mockUpdateDeviceType).toHaveBeenCalledWith({
      fetchPolicy: "no-cache",
      variables: {
        internalId: "abc3",
        name: "Covalent Observer",
        loincCode: "1234-3",
        manufacturer: "Vitamin Tox LLC",
        model: "Model CD",
        swabTypes: ["789"],
      },
    });

    expect(await screen.findByText("Redirected to /admin")).toBeInTheDocument();
  });
});
