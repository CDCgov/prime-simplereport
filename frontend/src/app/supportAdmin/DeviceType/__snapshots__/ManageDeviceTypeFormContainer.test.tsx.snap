// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManageDeviceTypeFormContainer renders the Manage Device Type Form Container item 1`] = `undefined`;
